### ISSUE TITLE
False warning with #[warn(unused_imports)] for std::ops::Deref

### ISSUE THREAD

**Issue by raphaelcohn** [rust-lang/rust#51556]
The lint `#[warn(unused_imports)]` produces a false warning of "warning: unused import: `::std::ops::Deref`", which is incorrect; following the warning and removing the import creates code that does not compile.

In my `lib.rs` I have `use ::std::ops::Deref`.

In submodules, which use `use super::*` imports, I have some code that explicitly calls `xxx.deref()`. Following the warning and removing the import from `lib.rs` causes code to break. (As an aside, this code explicitly calls deref() because trying to auto-deref confuses the compiler in this code).

**Comment by ishitatsuyuki** [comment]
Please show a full example of the code that breaks. Also, if auto deref is an issue for you, please use [UFCS](https://doc.rust-lang.org/book/second-edition/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation-calling-methods-with-the-same-name).

**Comment by ExpHP** [comment]
Sounds like #45268

**Comment by raphaelcohn** [comment]
@ExpHP - Yep, it sounds very similar, and one would hope a fix to that would fix this. I also use a lot of tiny submodules for similar reasons - and lots of `include!()` so I can have one struct, const or function per file - makes it far, far easier to work on large code bases with many devs...

**Comment by raphaelcohn** [comment]
@ExpHP Anyone on the compiler team made any progress on this? As a fault, similar things are cropping up quite often in my code... It's becoming a code quality issue for me.

**Comment by ExpHP** [comment]
Heh; I'm also just a user!  I know how you feel about it, because it continues to crop increasingly often in my code as well.  That said, I am subscribed to updates in all of the duplicates I could find, and haven't heard much activity.  (though two organization members were [recently hit by it](https://github.com/rust-lang/rust/issues/53682))

Presumably anything not related to the 2018 edition is lower priority right now, given its impending release.  This doesn't mean things can't move forward on this, but I suspect it might take a PR from a non-team member to push things forward.

I tried looking into it a long time ago, but the best I could manage to come up with was this [amateur analysis here](https://github.com/rust-lang/rust/issues/45268#issuecomment-341936496) with no idea how to fix it. :/

**Comment by ExpHP** [comment]
Uhh, wow! Certainly sounds like you needed to vent!

A number of the things you point out are things that are being actively worked on, and are an awful lot more complicated than they may seem on the surface.

The deal with associated consts---which I assume is the fact that they don't work as the `N` in `[T; N]` in generic code---is well known and it bothers the heck out of *everybody;* but it's not simply that this one little case wasn't accounted for; rather, I imagine the case is that all of the other places where consts work in `[T; N]` are a pure *hack!* Integers are not part of the type system in rust.  Type inference, unification and trait solving know nothing about them.  Making this one last thing work in a scalable manner would seem to require basically half of the entire infrastructure of const generics... which is a planned feature that [recently got a big PR.](https://github.com/rust-lang/rust/pull/53645)

(yeah, I know, the current error message sucks.  I chalk it up to whatever hack there must be to make everything else work)

Other things like `TryFrom` are blocked on yet other things (`never_type`) in order to make sure they *don't* come out unpolished.

**Comment by raphaelcohn** [comment]
@ExpHP ta for the detailed reply. Signing off for now, least I side track this issue with irrelevancy...

**Comment by sanxiyn** [comment]
I am so sorry that this is causing trouble for you. On the other hand, @ExpHP's analysis is correct. Closing as a duplicate.

**Comment by raphaelcohn** [comment]
@sanxiyn Good practice would be to put a reference to the duplicate bug in your close message...

**Comment by raphaelcohn** [comment]
Ta
